### DISCUSSION TITLE
Periodic sync example?

### DISCUSSION THREAD

**Discussion by thytrherfvsefasfd443sf** [rmartin16/qbittorrent-api#376] [Q&A] [answered]
I saw [this page](https://qbittorrent-api.readthedocs.io/en/latest/apidoc/sync.html), but I cannot find example code. Is the following the best way to do periodic sync (continuously displaying current information)? I am wondering if there this library is providing a better way like internally caching and merging the "maindata", so that I don't have to maintain rid and previous values.

```python
rid = 0
us = -1
ds = -1
while True:
    maindata = qbt_client.sync_maindata(rid=rid)
    rid = maindata["rid"]
    ss = maindata["server_state"]

    us = ss.get("up_info_speed", us)
    ds = ss.get("dl_info_speed", ds)
    print(f"Upload: {us}, Download: {ds}")
    time.sleep(2)
```

**Comment by rmartin16** [comment, marked as answer]
oh yeah....the docs don't include the [simplest](https://github.com/rmartin16/qbittorrent-api/blob/27c537c03ae2ceaf086fb88d66ac2b8b42f7fdc6/src/qbittorrentapi/sync.py#L120) way to use sync data. I should add a better example.

Give this a go:
```python
import time
from collections.abc import Mapping

def update(d, u):
    for k, v in u.items():
        d[k] = update(d.get(k, {}), v) if isinstance(v, Mapping) else v
    return d

maindata = client.sync_maindata()

while True:
    print(f"Upload: {maindata.server_state.up_info_speed}, Download: {maindata.server_state.dl_info_speed}")
    maindata = update(maindata, client.sync.maindata.delta())
    time.sleep(1)
```

`client.sync.maindata.delta()` can keep track of the `rid` for you between subsequent calls.

Although, there isn't anything to keep a full up-to-date `maindata`....that probably makes sense to add, though. I can probably just add a method to the return object from `client.sync_maindata()` to sync it.
